docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,392 @@
+
+Continued example: Dealing with untyped code
+--------------------------------------------
+
+Suppose we want the model from above to detect the sink in this snippet:
+
+.. code-block:: js
+
+  import { getConnection } from "@example/db";
+  let connection = getConnection();
+  connection.query(q); // <-- add 'q' as a SQL injection sink
+
+There is no type annotation on **connection**, and there is no indication of what **getConnection()** returns.
+Using a **typeModel** tuple we can tell our model that this function returns an instance of **mysql.Connection**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: typeModel
+      data:
+        - ["mysql.Connection", "@example/db", "Member[getConnection].ReturnValue"]
+
+To break this down:
+
+- Since we're providing type information, we add a tuple to the **typeModel** extension point.
+- The first column, **"mysql.Connection"**, names the type that we're adding a new definition for.
+- The second column, **"@example/db"**, begins the search at imports of the hypothetical NPM package **@example/db**.
+- **Member[getConnection]** selects references to the **getConnection** member from that package.
+- **ReturnValue** selects the return value from a call to that member.
+
+The new model states that the return value of **getConnection()** has type **mysql.Connection**.
+Combining this with the sink model we added earlier, the sink in the example is detected by the model.
+
+The mechanism used here is how library models work for both TypeScript and plain JavaScript.
+A good library model contains **typeModel** tuples to ensure it works even in codebases without type annotations.
+For example, the **mysql** model that is included with the CodeQL JS analysis includes this type definition (among many others):
+
+.. code-block:: yaml
+
+  - ["mysql.Connection", "mysql", "Member[createConnection].ReturnValue"]
+
+Reference
+=========
+
+Extension points
+----------------
+
+sourceModel(type, path, kind)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adds a new taint source. Most taint-tracking queries will use the new source.
+
+- **type**: Name of a type from which to evaluate **path**.
+- **path**: Access path leading to the source.
+- **kind**: Kind of source to add. Currently only **remote-flow** is used.
+
+Example:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sourceModel
+      data:
+        - ["global", "Member[user].Member[name]", "remote-flow"]
+
+sinkModel(type, path, kind)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adds a new taint sink. Sinks are query-specific and will usually affect one or two queries.
+
+- **type**: Name of a type from which to evaluate **path**.
+- **path**: Access path leading to the sink.
+- **kind**: Kind of sink to add. See `sink kinds <#sink-kinds>`_ for a list of supported kinds.
```

Comment:
The `sink-kind` link doesn't work (the relevant header has `id="id3"`).  
But I suppose that will resolve with the other link issue.

Reply:
As mentioned earlier, removed the link